blog: using mongodb with node-red (#1900)
FlowFuse/website: src/blog/2024/04/using-mongodb-with-node-red.md

---

> title: Using Mongodb With Node-RED

Reviewer: ```suggestion
title: Using MongoDB With Node-RED
```

> subtitle: Guide to integrating MongoDB database into Node-RED Applications, for retrieving, updating, and deleting data

Reviewer: ```suggestion
subtitle: Guide to integrating MongoDB database into Node-RED Applications, for writing, updating, and deleting data.
```

description: Learn how to seamlessly integrate MongoDB, a NoSQL database, into your Node-RED applications with this step-by-step guide. Gain a fundamental understanding of MongoDB's key concepts and terminologies. 

> date: 2024-04-05

Reviewer: ```suggestion
date: 2024-04-12
```

authors: ["sumit-shinde"]

> image: 

Reviewer: Art request please :)

Reviewer: @sumitshinde-84 Can you link the issue in this thread?

tags:
    - posts
    - node-red
    - MongoDB
    - NoSQL
---

MongoDB, a popular NoSQL database, is a favored choice among developers. As MongoDB continuously releases updates to enhance efficiency and suitability for manufacturing purposes, we find it essential to provide content for our Node-RED users. In this guide, I will cover everything from MongoDB foundations to setup, providing you with step-by-step instructions for integration into your Node-RED applications.

<!--more-->

## What is MongoDB

> MongoDB is an open-source, NoSQL database system designed for flexibility and scalability in handling data for modern applications. It stores data in a format called JSON-like documents, which offers a schema-less approach where each document can have its unique structure. This flexibility allows for easy schema evolution and dynamic querying. MongoDB is widely used in web and mobile applications for its ability to handle large volumes of data and support real-time applications efficiently.

Reviewer: ```suggestion
MongoDB is an open-source, NoSQL database system designed for flexibility and scalability in handling data for modern applications. It stores data in documents, which are JSON-like structures. MongoDB offers a schema-less approach where each document can have its unique structure. This flexibility allows for easy schema evolution and dynamic querying. MongoDB is widely used in web and mobile applications for its ability to handle large volumes of data and support real-time applications efficiently.
```

In MongoDB, instead of tables, rows, and columns as in SQL databases, we have collections, documents, and fields.

> !["Annotomy of MongoDB document"](./images/using-mongodb-with-node-red-annotomy-of-mongodb-document.png "Annotomy of MongoDB document"){data-zoomable}

Author: @ZJvandeWeg Should we recreate this graphic with our ff colors? I downloaded it from the internet.

Reviewer: Please add it to the art request I mentioned earlier

- MongoDB a fully managed cloud database service, provides tools like Atlas Data Explorer, Real-Time Performance Metrics, Atlas Search, and Atlas Data Lake. These simplify database management with automated backups, point-in-time recovery, security features, and seamless scalability, which enhances operational efficiency for developers and administrators. Additionally, MongoDB has released Atlas specialized for manufacturing. For more information, refer to  [MongoDB Launches Atlas for Manufacturing and Automotive](https://www.mongodb.com/company/newsroom/press-releases/mongodb-atlas-for-manufacturing-and-automotive)
- More Importantly, MongoDB's native support for JSON documents simplifies data storage and retrieval, making it perfect for integration with Node-RED.

## Setting Up MongoDB with Node-RED
To kickstart the MongoDB integration with Node-RED, let's begin by installing the MongoDB custom node and familiarizing ourselves with MongoDB configuration details:

### Installing MongoDB custom node

- Click the Node-RED Settings (top-right)
- Click "Manage Palette"
- Switch to the "Install" tab
- Search for `node-red-contrib-mongodb4`

### Understanding MongoDB Configuration

> To establish a connection between our Node-RED application and the MongoDB database, it's essential to have the following configuration details prepared ( While writing the article, I have utilized MongoDB Atlas, but feel free to choose any MongoDB setup, as the instructions in the article are adaptable for any setup ) :

Reviewer: ```suggestion
While writing the article, I have utilized MongoDB Atlas, but feel free to choose any MongoDB setup, as the instructions in the article are adaptable for any setup. To establish a connection between our Node-RED application and the MongoDB database, it's essential to have the following configuration details prepared:
```

> - Host: IP address or hostname of your MongoDB server.

Reviewer: ```suggestion
- `Host`: IP address or hostname of your MongoDB server.
```

> - Port: By default, MongoDB uses port 27017. ( If you're using a managed MongoDB service like MongoDB Atlas, this information may not be required.)

Reviewer: ```suggestion
- `Port`: By default, MongoDB uses port 27017. ( If you're using a managed MongoDB service like MongoDB Atlas, this information may not be required.)
```

> - Database: The name of the MongoDB database you want to connect to.
> - User: Username with the necessary privileges to access the specified database.
> - Password: Corresponding password for the username.

Reviewer: ```suggestion
- `Database`: The name of the MongoDB database you want to connect to.
- `User`: Username with the necessary privileges to access the specified database.
- `Password`: Corresponding password for the username.
```

- Click on the add variable button and “add variables” for each of the configuration data that we discussed in the above section.
- Click on the save button and restart the instance by clicking on the top right Action button and selecting the restart option.

### Configuring MongoDB node

Now let's configure MongoDB4 node using added environment variables 

- Drag the MongoDB4 node onto the canvas.
- Click on the node to select it, Then, click on the edit icon next to the connection input field.
- Add your environment variables as shown in the image below.

!["Screenshot displaying connection configuration of MongoDB 4 node."](./images/using-mongodb-with-node-red-mongodb-node-connection-configuration.png "Screenshot displaying connection configuration of MongoDB 4 node."){data-zoomable}

## Using MongoDB with Node-RED

In this section, we'll delve into leveraging MongoDB to perform operations such as storing, retrieving, updating, and deleting data within Node-RED. To enhance understanding and engagement, we'll develop a basic customer relationship manager system utilizing MongoDB and Node-RED. 

For the CRM system we will develop, the data structure will be as follows:

> ```json

Reviewer: I think you need some `{% raw %}` tags?

- Drag the ui-table widget onto the canvas, create a new ui-group for it, and set max rows according to your preference.

!["Screenshot displaying ui-table widget configuration"](./images/using-mongodb-with-node-red-table-widget.png "Screenshot displaying ui-table widget configuration"){data-zoomable}

- Connect the nodes' wires as shown in the below image.

!["Screenshot displaying connections of wires in the 'Retrive Data from Database' flow"](./images/using-mongodb-with-node-red-mongodb-find-flow.png "Annotomy of MongoDB document"){data-zoomable}

### Updating Customer Data to Database

- Drag the ui-form widget onto the canvas, create a new ui-group for it, and add form elements for "id" and "status".

!["Screenshot displaying form widget configuration to update data in MongoDB"](./images/using-mongodb-with-node-red-insert-data-form.png "Screenshot displaying form widget configuration to update data in MongoDB"]{data-zoomable}

- Drag the change node onto the canvas. In the change node, set `msg.payload` as the following JSON object:

```json
[

>     // Specifying the criteria to identify the document to update, based on its _id field

Reviewer: Not sure if we should allow people to copy+paste illegal JSON?

Author: How about if we add the following note before code blocks:
Note:- *Please note that the comments provided are for explanation purposes only and should not be copied along with the code*

Reviewer: Not as a comment in the code I think, just above it?

- Drag the MongoDB4 node onto canvas, make sure to select the connection, and enter “updateOne” into the operation input field.

!["Screenshot displaying configuration of MongoDB 4 node for updating data"](./images/using-mongodb-with-node-red-mongodb-update-node-configuration.png "Screenshot displaying configuration of MongoDB 4 node for updating data"){data-zoomable}

- Connect the nodes' wires as shown in the below image.

!["Screenshot displaying connections of wires in the 'Update Data from Database' flow"](./images/using-mongodb-with-node-red-mongodb-updateone-flow.png "Screenshot displaying connections of wires in the 'Update Data from Database' flow"){data-zoomable}

### Deleting Customer Data to Database

- Drag the ui-form widget onto the canvas, create a new ui-group for it, and add form elements for id and name.

!["Screenshot displaying form widget configuration to delete data in MongoDB"](./images/using-mongodb-with-node-red-delete-data-form.png "Screenshot displaying form widget configuration to delete data in MongoDB"){data-zoomable}

- Drag the change node onto the canvas. In the change node, set `msg.payload` as the following JSON object:

> ```json

Reviewer: As YAML is a superset of JSON, we can fix the highlighting with:

```suggestion
```yaml
```

- Drag the MongoDB4 node onto canvas, make sure you have selected the added connection, Enter the collection name that you want to drop from the database, and enter “drop” into the operation input field.

!["Screenshot displaying configuration of MongoDB4 node for droping collection from database"](./images/using-mongodb-with-node-red-mongodb-drop-node-configuration.png "Screenshot displaying configuration of MongoDB4 node for droping collection from database"){data-zoomable}

- Connect the nodes' wires as shown in the below image.

!["Screenshot displaying connections of wires in the 'Drop collecton from Database' flow"](./images/using-mongodb-with-node-red-mongodb-drop-flow.png "Screenshot displaying connections of wires in the 'Drop collecton from Database' flow"){data-zoomable}

## Debugging the operations

To debug MongoDB operations in Node-RED, simply add a debug node after the MongoDB4 node. This enables you to monitor the execution of operations and diagnose any errors that may occur. Below are sample debug messages indicating successful operations for each operation we have covered in this guide.

> ```

Reviewer: ```suggestion
```js
```

- With your flow updated to include the above, click the "Deploy" button in the top-right of the Node-RED Editor.
- Locate the 'Open Dashboard' button at the top-right corner of the Dashboard 2.0 sidebar and click on it to navigate to the dashboard.

Now To drop tables, and retrieve table data, click on the 'Inject Node' button. For customer data insertion, updates, and deletions, navigate to fill in the corresponding forms and submit.

!["Screenshot displaying dashboard view of CRM System"](./images/using-mongodb-with-node-red-crm-system-node-red-dashboard-view.png "Screenshot displaying dashboard view of CRM System"){data-zoomable}

## Additional resources

This section contains helpful links to other content. It isn’t required, so consider it supplemental.

- [MongoDB best practices](https://www.mongodb.com/basics/best-practices) for to increase performance
- [Ebook: MongoDB in manufacturing](https://www.mongodb.com/collateral/manufacturing-with-mongodb).
- [Manufacturing & Automotive Industry Knowledge Accelerator](https://learn.mongodb.com/learning-paths/manufacturing-automotive-industry-knowledge-accelerator)

## Conclusion

> In this guide, we've covered the integration of MongoDB with Node-RED to create dynamic applications. We've walked through the foundation, setup process, demonstrated how to perform CRUD operations, and discussed debugging techniques. Additionally, to enhance understanding and engagement we've built a Customer Relationship Management system.

Reviewer: Should we add a link to the blueprint that's available too?

Author: Yes, let's include the link to the blueprint.

    "upsertedCount": 0,
    "matchedCount": 1 // Indicates one document is matched for the provided query
}

// Message received after delete operation successful
{
    "acknowledged": true,
    "deletedCount": 1 // Indicates that one document is deleted
}

// To indicate collection is dropped successfully and true value will be returned
```

### Deploying the flow

!["Screenshot displaying flow of CRM System"](./images/using-mongodb-with-node-red-crm-system-node-red-flow.png "Screenshot displaying flow of CRM System"){data-zoomable}

- With your flow updated to include the above, click the "Deploy" button in the top-right of the Node-RED Editor.
- Locate the 'Open Dashboard' button at the top-right corner of the Dashboard 2.0 sidebar and click on it to navigate to the dashboard.

Now To drop tables, and retrieve table data, click on the 'Inject Node' button. For customer data insertion, updates, and deletions, navigate to fill in the corresponding forms and submit.

!["Screenshot displaying dashboard view of CRM System"](./images/using-mongodb-with-node-red-crm-system-node-red-dashboard-view.png "Screenshot displaying dashboard view of CRM System"){data-zoomable}

> ## Additional resources

Author: What do you think about this new section? Should we add it to our next upcoming post as well?
